lib/ACL/ACLManager.php

```diff
@@ -168,12 +169,32 @@ public function getPermissionsForPathFromRules(string $path, array $rules): int
 	 * @return int
 	 */
 	private function calculatePermissionsForPath(array $rules): int {
-		// first combine all rules with the same path, then apply them on top of the current permissions
-		// since $rules is sorted parent first rules for subfolders overwrite the rules from the parent
-		return array_reduce($rules, function (int $permissions, array $rules): int {
-			$mergedRule = Rule::mergeRules($rules);
-			return $mergedRule->applyPermissions($permissions);
-		}, Constants::PERMISSION_ALL);
+		if ($this->inheritMergePerUser) {
+			// first combine all rules for the same user-mapping by path order
+			// then merge the results with allow overwrites deny
+			$rulesPerMapping = [];
+			foreach ($rules as $rulesForPath) {
+				foreach ($rulesForPath as $rule) {
+					$mapping = $rule->getUserMapping();
+					$key = $mapping->getType() . '/' . $mapping->getId();
+					if (!isset($rulesPerMapping[$key])) {
+						$rulesPerMapping[$key] = Rule::defaultRule();
+					}
+
+					$rulesPerMapping[$key]->applyRule($rule);
+				}
+			}
+
+			$mergedRule = Rule::mergeRules($rulesPerMapping);
```

Comment:
Here we would only have something like:

```php
$rulesPerMapping = [
    'group/gid1' => Rule,
    'group/gid2' => Rule,
    'user/uid' => Rule,
];
```

Where uid is the user id of the current user, and where there can be multiple gid depending on the user's groups, right?

Reply:
yes